src/streamlit_extras/gallery/__init__.py

```diff
@@ -0,0 +1,299 @@
+    /* transition: 0.3s; */
+    flex-basis: 0;
+    flex-grow: 1;
+    border-color: rgba(49, 51, 63, 0.2);
+    border-radius: 12px;
+    border-style: solid;
+    border-width: 1px;
+    }
+
+    /* On mouse-over, add a deeper shadow
+    .card:hover {
+    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
+    }*/
+
+    /* Add some padding inside the card container */
+    .card-container {
+    padding: 20px 16px;
+
+    }
+
+    /* Add rounded corners to the top left and the top right corner of the image */
+    .card-img {
+    border-radius: 12px 12px 0 0;
+    height: 150px;
+    object-fit: cover;
+    }
+</style>
+"""
+
+
+@st.experimental_memo
+def _card_builder(
+    df_element,
+    columns,
+) -> str:
+
+    # Validate card types
+    assert isinstance(columns, dict), NotImplementedError(
+        "For now, column types must be explicitly given."
+    )
+
+    # Check there's one image only
+    image_type_columns = [c for c in columns if columns[c]["type"] == "image"]
+    assert (
+        len([c for c in columns if columns[c]["type"] == "image"]) == 1
+    ), "There must be exactly one type 'image'."
+    image_type_column = image_type_columns[0]
+
+    # Check there's one title only
+    title_type_columns = [c for c in columns if columns[c]["type"] == "title"]
+    assert (
+        len([c for c in columns if columns[c]["type"] == "title"]) == 1
+    ), "There must be exactly one type 'title'."
+    title_type_column = title_type_columns[0]
+
+    tag_type_columns = [c for c in columns if columns[c]["type"] == "tag"]
+    text_type_columns = [c for c in columns if columns[c]["type"] == "text"]
+    html_type_columns = [c for c in columns if columns[c]["type"] == "html"]
+
+    # Create card
+    card = div(_class="card")(
+        img(
+            src=getattr(df_element, image_type_column),
+            _class="card-img",
+            style=styles(width="100%"),
+        ),
+        div(_class="card-container")(
+            # Title
+            p(
+                getattr(df_element, title_type_column),
+                style=styles(
+                    font_size="22px",
+                    margin_top="5px",
+                    margin_bottom="0px",
+                    font_weight="bold",
+                ),
+            ),
+            # Tags
+            "&nbsp;&nbsp;".join(
+                [_to_tag(getattr(df_element, column)) for column in tag_type_columns]
+            ),
+            p(),
+            # Text
+            # Look at https://stackoverflow.com/questions/55351096/ellipsis-three-dots-expand-and-collapse-the-text
+            # To add '...' logic.
+            "\n".join(
+                [str(p(getattr(df_element, column))) for column in text_type_columns]
+            ),
+            # Links
+            "<br>".join([getattr(df_element, column) for column in html_type_columns]),
+        ),
+    )
+    return str(card) + _CARD_CSS
+
+
+@extra
+@st.experimental_memo
```

Comment:
Does this actually make a noticeable difference? Is it fairly slow to build? It will mean that it doesn't work well with older versions of streamlit that don't support memoizing st outputs

Reply:
That's a good point. Maybe there's a safe way to try/except it so it works in both cases... I'll check locally if it really brings value. I kind of remember from your past analyses that not really, though